crypto/src/key/secp256k1/extended_keys.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::key::hdkd::chain_code::CHAINCODE_LENGTH;
+use crate::key::hdkd::{
+    chain_code::ChainCode,
+    child_number::ChildNumber,
+    derivable::{Derivable, DerivationError},
+};
+use crate::key::secp256k1::{Secp256k1PrivateKey, Secp256k1PublicKey};
+use crate::random::{CryptoRng, Rng};
+use hmac::{Hmac, Mac};
+use secp256k1;
+use secp256k1::SECP256K1;
+use serialization::{Decode, Encode};
+use sha2::Sha512;
+use zeroize::Zeroize;
+
+// Create alias for HMAC-SHA512
+type HmacSha512 = Hmac<Sha512>;
+
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
+pub struct Secp256k1ExtendedPrivateKey {
+    /// Chain code
+    pub chain_code: ChainCode,
+    /// Private key
+    pub private_key: Secp256k1PrivateKey,
+}
+
+impl Secp256k1ExtendedPrivateKey {
+    pub fn new_master(seed: &[u8]) -> Result<Secp256k1ExtendedPrivateKey, DerivationError> {
+        let mut mac =
+            HmacSha512::new_from_slice(b"Bitcoin seed").expect("HMAC can take key of any size");
+        mac.update(seed);
+        let mut result = mac.finalize().into_bytes();
+
+        let private_key = secp256k1::SecretKey::from_slice(&result[..32])
+            .map_err(|_| DerivationError::KeyDerivationError)?
+            .into();
+
+        let chain_code: [u8; CHAINCODE_LENGTH] =
+            result[32..].try_into().expect("Chaincode size is 32 bytes");
+        let chain_code = ChainCode::from(chain_code);
```

Comment:
You can use the `.split()` method on `GenericArray` to extract the key part and the chain code part with compile-time bounds checking.

Something like this (may need adjustment):
```rust
use generic_array::{GenericArray, typenum::U32, sequence::Split};
type ByteArray32 = GenericArray<u8, U32>;
let (private_key, chian_code): (ByteArray32, ByteArray32) = result.split();
```

The `ByteArray32` type is going to be convertible to `[u8; 32]` and `&[u8]`.